distributed/diagnostics/nvml.py

```diff
@@ -1,19 +1,30 @@
 import pynvml
 
-pynvml.nvmlInit()
-count = pynvml.nvmlDeviceGetCount()
+pynvml_initialized = False
+handles = None
 
-handles = [pynvml.nvmlDeviceGetHandleByIndex(i) for i in range(count)]
+
+def _ensure_pynvml_initialized():
+    global pynvml_initialized, handles
+    if not pynvml_initialized:
+        pynvml.nvmlInit()
+        count = pynvml.nvmlDeviceGetCount()
+        handles = [pynvml.nvmlDeviceGetHandleByIndex(i) for i in range(count)]
+        pynvml_initialized = True
 
 
 def real_time():
+    global handles
+    _ensure_pynvml_initialized()
```

Comment:
I recommend that we remove the global here, and have the function return handles.  This again should hide some of the jankiness of the global state inside of the function.

Maybe

```python
handles = pynvml_handles()  # both inits if necessary, and returns handles
```

Reply:
I don't want to hold up getting this merged, but here's another suggestion (untested)
```
class PynvmlMgr:
    def __init__(self):
        self.__initialized = False
        self.__handles = []

    @property
    def handles(self):
        self.__ensureInitialized()
        return self.__handles

    def __ensureInitialized(self):
        if not self.__initialized:
            pynvml.nvmlInit()
            count = pynvml.nvmlDeviceGetCount()
            self.__handles = [pynvml.nvmlDeviceGetHandleByIndex(i) for i in range(count)]
            self.__initialized = True

pynvmlMgr = PynvmlMgr()

# Then just use handles:
pynvmlMgr.handles
```

Reply:
Makes sense to me - made the change.